Source/WebCore/platform/graphics/ComplexTextController.cpp

```diff
@@ -757,7 +757,8 @@ void ComplexTextController::adjustGlyphsAndAdvances()
 
                 unsigned characterIndexInRun = characterIndex + complexTextRun.stringLocation();
                 bool isFirstCharacter = !(characterIndex + complexTextRun.stringLocation());
-                bool isLastCharacter = characterIndexInRun + 1 == m_run.length() || (U16_IS_LEAD(character) && characterIndexInRun + 2 == m_run.length() && U16_IS_TRAIL(charactersSpan[characterIndex + 1]));
+                bool isLastCharacter = characterIndexInRun + 1 == m_run.length()
+                    || (U16_IS_LEAD(character) && characterIndexInRun + 2 == m_run.length() && characterIndex + 2 == charactersSpan.size() && U16_IS_TRAIL(charactersSpan[characterIndex + 1]));
```

Comment:
Why `characterIndex + 2 == charactersSpan.size()` rather than `characterIndex + 2 >= charactersSpan.size()`?

Reply:
`>=` is also fine. This condition is checking if this is the last character of m_run. m_run is split into some m_complexTextRuns. For example, assuming m_run "abc😀" is split into "abc" and "😀". If "😀" is the last character of m_run, both conditions `characterIndexInRun + 2 == m_run.length()` and `characterIndex + 2 == charactersSpan.size()` are true.

Reply:
There’s a guarantee that no characters are after `m_run` in `charactersSpan`?

Reply:
I don't understand your question. For example, assuming m_run "abc😀😎xyz" is split into "abc", "😀😎" and "xyz". And, "😎" is the current character, this is not the last character because "xyz" follows. The condition `characterIndexInRun + 2 == m_run.length()` is false.

Reply:
Never mind. I thought that maybe `m_run` was sometimes not at the very end of the span but was the "last" but I guess that is not the case.